pyvista/utilities/algorithms.py

```diff
@@ -0,0 +1,271 @@
+            out = self.GetOutputData(outInfo, 0)
+            output = inp.cast_to_polydata(deep=False)
+            out.ShallowCopy(output)
+        except Exception as e:  # pragma: no cover
+            traceback.print_exc()
+            raise e
+        return 1
+
+
+class AddIDsAlgorithm(PreserveTypeAlgorithmBase):
+    """Internal helper algorithm to add point or cell IDs."""
+
+    def __init__(self, point_ids=True, cell_ids=True):
+        """Initialize algorithm."""
+        super().__init__()
+        if not point_ids and not cell_ids:  # pragma: no cover
+            raise ValueError('IDs must be set for points or cells or both.')
+        self.point_ids = point_ids
+        self.cell_ids = cell_ids
+
+    def RequestData(self, request, inInfo, outInfo):
+        """Perform algorithm execution."""
+        try:
+            inp = wrap(self.GetInputData(inInfo, 0, 0))
+            out = self.GetOutputData(outInfo, 0)
+            output = inp.copy()
+            if self.point_ids:
+                output.point_data['point_ids'] = np.arange(0, output.n_points, dtype=int)
+            if self.cell_ids:
+                output.cell_data['cell_ids'] = np.arange(0, output.n_cells, dtype=int)
+            if output.active_scalars_name in ['point_ids', 'cell_ids']:
+                output.active_scalars_name = inp.active_scalars_name
+            out.ShallowCopy(output)
+        except Exception as e:  # pragma: no cover
+            traceback.print_exc()
+            raise e
+        return 1
+
+
+class CrinkleAlgorithm(_vtk.VTKPythonAlgorithmBase):
+    """Internal helper algorithm to crinkle cell IDs."""
+
+    def __init__(self):
+        """Initialize algorithm."""
+        super().__init__(
+            nInputPorts=2,
+            outputType='vtkUnstructuredGrid',
+        )
+
+    def RequestData(self, request, inInfo, outInfo):
+        """Perform algorithm execution."""
+        try:
+            clipped = wrap(self.GetInputData(inInfo, 0, 0))
+            source = wrap(self.GetInputData(inInfo, 1, 0))
+            out = self.GetOutputData(outInfo, 0)
+            output = source.extract_cells(np.unique(clipped.cell_data['cell_ids']))
+            out.ShallowCopy(output)
+        except Exception as e:  # pragma: no cover
+            traceback.print_exc()
+            raise e
+        return 1
+
+
+def outline_algorithm(inp, generate_faces=False):
+    """Add vtkOutlineFilter to pipeline."""
+    alg = _vtk.vtkOutlineFilter()
+    set_algorithm_input(alg, inp)
+    alg.SetGenerateFaces(generate_faces)
+    return alg
+
+
+def extract_surface_algorithm(inp, pass_pointid=False, pass_cellid=False, nonlinear_subdivision=1):
+    """Add vtkDataSetSurfaceFilter to pipeline."""
+    surf_filter = _vtk.vtkDataSetSurfaceFilter()
+    surf_filter.SetPassThroughPointIds(pass_pointid)
+    surf_filter.SetPassThroughCellIds(pass_cellid)
+    if nonlinear_subdivision != 1:
+        surf_filter.SetNonlinearSubdivisionLevel(nonlinear_subdivision)
```

Comment:
I suppose. I copy/pasted this logic from the existing filter:

https://github.com/pyvista/pyvista/blob/4fe5620bfd3051e500e05960ea1391afd63a59d9/pyvista/core/filters/data_set.py#L4433-L4434

Reply:
> Or is this function again meant to be used internally, with us passing in 1 most of the time?

Generally yes. Since I copied/pasted this logic from elsewhere and this is mostly meant to be used internally, I am going to leave this as is for now.